### PR TITLE
Refactor the storage function module

### LINKED ISSUE #288: Refactor the storage function module
### Feature request:

refactor the storage function module & remove unused code
Support storage manager using plug-in mode
### Use case:
{
				Name:   hybrid.StoragePattern,
				Enable: false,
				Config: &storage.Config{
					GCInitialDelay: 0 * time.Second,
					GCInterval:     15 * time.Second,
					DriverConfigs: []*storage.DriverConfig{
						{
							Name: local.DiskDriverName,
							GCConfig: &storage.GCConfig{
								YoungGCThreshold:  100 * unit.GB,
								FullGCThreshold:   5 * unit.GB,
								CleanRatio:        3,
								IntervalThreshold: 2 * time.Hour,
							},
						}, {
							Name: local.MemoryDriverName,
							GCConfig: &storage.GCConfig{
								YoungGCThreshold:  100 * unit.GB,
								FullGCThreshold:   5 * unit.GB,
								CleanRatio:        3,
								IntervalThreshold: 2 * time.Hour,
							},
						},
					},
				},

### PR DESCRIPTION
**Pull request by 244372610**
## Description
refactor the storage function module & remove unused code
Support storage manager using plug-in mode

fix #288 
## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Documentation Update (if none of the other choices apply)

## Checklist:

- [x] Code compiles correctly.
- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have read the **CONTRIBUTING** document.
- [ ] I have added tests to cover my changes.
- [ ] All new and existing tests passed.

### CODE REVIEW

File: cdnsystem/config/path_darwin.go [hunk 1]
```diff
@@ -0,0 +1,23 @@
...
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package config
+
+const (
+	DefaultDiskBaseDir = "/tmp/cdnsystem"
```
Comment (gaius-qi) [inline cdnsystem/config/path_darwin.go:20]: `Linux`and `Darwin` should use the same type of directory, if you use the tmp directory, please add the tmp directory in `internal/dfpath`.

File: cdnsystem/config/path_darwin.go [hunk 2]
```diff
@@ -0,0 +1,23 @@
...
+
+	DefaultMemoryBaseDir = "/tmp/memory/dragonfly"
```
Comment (gaius-qi) [inline cdnsystem/config/path_darwin.go:22]: `Linux`and `Darwin` should use the same type of directory, if you use the tmp directory, please add the tmp directory in `internal/dfpath`.

File: cdnsystem/config/path_linux.go [hunk 3]
```diff
@@ -0,0 +1,23 @@
...
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package config
+
+const (
+	DefaultDiskBaseDir = "/home/admin/ftp"
```
Comment (gaius-qi) [inline cdnsystem/config/path_linux.go:20]: `Linux`and `Darwin` should use the same type of directory, if you use the tmp directory, please add the tmp directory in `internal/dfpath`.

File: cdnsystem/config/path_linux.go [hunk 4]
```diff
@@ -0,0 +1,23 @@
...
+
+	DefaultMemoryBaseDir = "/dev/shm/dragonfly"
```
Comment (gaius-qi) [inline cdnsystem/config/path_linux.go:22]: `Linux`and `Darwin` should use the same type of directory, if you use the tmp directory, please add the tmp directory in `internal/dfpath`.

File: cdnsystem/config/testdata/config/cdn.yaml [hunk 5]
```diff
@@ -0,0 +1,96 @@
+# This file is the template of cdn system configuration file.
```
Comment (gaius-qi) [inline cdnsystem/config/testdata/config/cdn.yaml:1]: Remove the comment.

File: cdnsystem/daemon/mgr/cdn/cache_writer_util.go [hunk 6]
```diff
@@ -95,7 +95,7 @@ func (cw *cacheWriter) writerPool(ctx context.Context, wg *sync.WaitGroup, write
 				go func(record *storage.PieceMetaRecord) {
 					defer wg.Done()
 					// todo 可以先塞入channel，然后启动单独goroutine顺序写文件
```
Comment (gaius-qi) [inline cdnsystem/daemon/mgr/cdn/cache_writer_util.go:97]: Comment in English

File: cdnsystem/daemon/mgr/cdn/cache_writer_util.go [hunk 7]
```diff
@@ -69,7 +69,7 @@ func (cw *cacheWriter) writerPool(ctx context.Context, wg *sync.WaitGroup, write
 				pieceLen := originPieceLen                 // 经过处理后写到存储介质的真实长度
```
Comment (gaius-qi) [inline cdnsystem/daemon/mgr/cdn/cache_writer_util.go:69]: Comment in English

File: cdnsystem/daemon/mgr/cdn/storage/disk/disk.go [hunk 8]
```diff
@@ -40,83 +40,76 @@ import (
 	"github.com/sirupsen/logrus"
 )
 
-const name = "disk"
+const StoragePattern = "disk"
```
Comment (gaius-qi) [inline cdnsystem/daemon/mgr/cdn/storage/disk/disk.go:43]: Rename `StoragePattern` to `StorageMode`.

File: cdnsystem/daemon/mgr/cdn/storage/storage_mgr.go [hunk 9]
```diff
@@ -93,6 +101,8 @@ type PieceMetaRecord struct {
 	PieceStyle  types.PieceFormat `json:"pieceStyle"`  // 1: PlainUnspecified
 }
 
+const fieldSeparator = ":"
```
Comment (gaius-qi) [inline cdnsystem/daemon/mgr/cdn/storage/storage_mgr.go:104]: Remove this variable.

File: cdnsystem/storedriver/driver.go [hunk 10]
```diff
@@ -115,10 +120,152 @@ type StorageInfo struct {
 	ModTime    time.Time // modified time
 }
 
-// GcConfig
-type GcConfig struct {
-	YoungGCThreshold  unit.Bytes    `yaml:"youngGCThreshold"`
-	FullGCThreshold   unit.Bytes    `yaml:"fullGCThreshold"`
-	CleanRatio        int           `yaml:"cleanRatio"`
-	IntervalThreshold time.Duration `yaml:"intervalThreshold"`
+func init() {
```
Comment (gaius-qi) [inline cdnsystem/storedriver/driver.go:123]: Don't use the init function arbitrarily.

File: cdnsystem/storedriver/local/local_driver.go [hunk 11]
```diff
@@ -17,119 +17,74 @@
...
-
-const MemoryStorageDriver = "memory"
+const (
+	DiskDriverName   = "disk"
+	MemoryDriverName = "memory"
+)
 
 var fileLocker = synclock.NewLockerPool()
 
 func init() {
```
Comment (gaius-qi) [inline cdnsystem/storedriver/local/local_driver.go:49]: Don't use the init function arbitrarily.

File: cdnsystem/storedriver/local/local_driver.go [hunk 12]
```diff
@@ -220,22 +189,31 @@ func (ds *diskStorage) Put(ctx context.Context, raw *storedriver.Raw, data io.Re
...
-		f, err = fileutils.OpenFile(path, os.O_WRONLY|os.O_APPEND|os.O_CREATE, 0644)
+		if f, err = fileutils.OpenFile(path, os.O_WRONLY|os.O_APPEND|os.O_CREATE, 0644); err != nil {
+			return err
+		}
 	} else {
-		f, err = fileutils.OpenFile(path, os.O_WRONLY|os.O_CREATE, 0644)
-	}
-	if err != nil {
-		return err
+		if f, err = fileutils.OpenFile(path, os.O_WRONLY|os.O_CREATE, 0644); err != nil {
```
Comment (gaius-qi) [inline cdnsystem/storedriver/local/local_driver.go:200]: Don't use fileutils library, use golang os library。

File: cdnsystem/storedriver/store_mgr.go [hunk 13]
```diff
@@ -18,34 +18,48 @@ package storedriver
...
+	var f = func(conf interface{}) (plugins.Plugin, error) {
+		cfg := &Config{}
+		if err := mapstructure.Decode(conf, cfg); err != nil {
+			return nil, fmt.Errorf("failed to parse config: %v", err)
+		}
+		// prepare the base dir
+		if !filepath.IsAbs(cfg.BaseDir) {
+			return nil, fmt.Errorf("not absolute path: %s", cfg.BaseDir)
+		}
+		if err := fileutils.MkdirAll(cfg.BaseDir); err != nil {
```
Comment (gaius-qi) [inline cdnsystem/storedriver/store_mgr.go:46]: Use `os.MkdirAll`.

**Review by gaius-qi — CHANGES_REQUESTED** [review]

File: cdnsystem/daemon/mgr/cdn/storage/storage_mgr.go (hunk 9 above)
Comment (244372610) [reply cdnsystem/daemon/mgr/cdn/storage/storage_mgr.go:104]: It is needed to assemble and split strings

File: cdnsystem/storedriver/store_mgr.go (hunk 13 above)
Comment (244372610) [reply cdnsystem/storedriver/store_mgr.go:46]: Made the default mode encapsulation

File: cdnsystem/storedriver/local/local_driver.go (hunk 12 above)
Comment (244372610) [reply cdnsystem/storedriver/local/local_driver.go:197]: Encapsulate the case that the file does not exist

Comment (gaius-qi) [reply cdnsystem/storedriver/local/local_driver.go:197]: It is not necessary to use this package.

File: cdnsystem/storedriver/store_mgr.go (hunk 13 above)
Comment (gaius-qi) [reply cdnsystem/storedriver/store_mgr.go:46]: It is not necessary to use this package.

**Review by gaius-qi — APPROVED** [review]

**Review by ansinlee — APPROVED** [review]

**Pull request merged**